sway/lock.c

```diff
@@ -0,0 +1,179 @@
+	wlr_session_lock_surface_v1_configure(surf->lock_surface, surf->output->width, surf->output->height);
+}
+
+static void handle_surface_destroy(struct wl_listener *listener, void *data) {
+	struct sway_session_lock_surface* surf = wl_container_of(listener, surf, destroy);
+	wl_list_remove(&surf->map.link);
+	wl_list_remove(&surf->destroy.link);
+	wl_list_remove(&surf->surface_commit.link);
+	wl_list_remove(&surf->output_mode.link);
+	output_damage_whole(surf->output);
+	free(surf);
+}
+
+static void handle_new_surface(struct wl_listener *listener, void *data) {
+	struct wlr_session_lock_surface_v1 *lock_surface = data;
+	struct sway_session_lock_surface* surf = calloc(1, sizeof(*surf));
+	if (surf == NULL)
+		return;
+
+	sway_log(SWAY_DEBUG, "new lock layer surface");
+
+	struct sway_output *output = lock_surface->output->data;
+	wlr_session_lock_surface_v1_configure(lock_surface, output->width, output->height);
+
+	surf->lock_surface = lock_surface;
+	surf->surface = lock_surface->surface;
+	surf->output = output;
+	surf->map.notify = handle_surface_map;
+	wl_signal_add(&lock_surface->events.map, &surf->map);
+	surf->destroy.notify = handle_surface_destroy;
+	wl_signal_add(&lock_surface->events.destroy, &surf->destroy);
+	surf->surface_commit.notify = handle_surface_commit;
+	wl_signal_add(&surf->surface->events.commit, &surf->surface_commit);
+	surf->output_mode.notify = handle_output_mode;
+	wl_signal_add(&output->wlr_output->events.mode, &surf->output_mode);
+}
+
+static void handle_unlock(struct wl_listener *listener, void *data) {
+	struct sway_session_lock_manager *lock_state = server.session_lock->data;
+
+	sway_log(SWAY_DEBUG, "session unlocked");
+	lock_state->locked = false;
+	lock_state->lock = NULL;
+
+	wl_list_remove(&lock_state->new_surface.link);
+	wl_list_remove(&lock_state->unlock.link);
+	wl_list_remove(&lock_state->abandon.link);
+
+	struct sway_seat *seat;
+	wl_list_for_each(seat, &server.input->seats, link) {
+		seat_set_exclusive_client(seat, NULL);
+		// copied from seat_set_focus_layer -- deduplicate?
+		struct sway_node *previous = seat_get_focus_inactive(seat, &root->node);
+		if (previous) {
+			// Hack to get seat to re-focus the return value of get_focus
+			seat_set_focus(seat, NULL);
+			seat_set_focus(seat, previous);
+		}
+	}
+
+
+	// redraw everything
+	for (int i = 0; i < root->outputs->length; ++i) {
+		struct sway_output *output = root->outputs->items[i];
+		output_damage_whole(output);
+	}
+}
+
+static void handle_abandon(struct wl_listener *listener, void *data) {
+	struct sway_session_lock_manager *lock_state = server.session_lock->data;
+
+	sway_log(SWAY_INFO, "session lock abandoned");
+	lock_state->lock = NULL;
+
+	wl_list_remove(&lock_state->new_surface.link);
+	wl_list_remove(&lock_state->unlock.link);
+	wl_list_remove(&lock_state->abandon.link);
+
+	struct sway_seat *seat;
+	wl_list_for_each(seat, &server.input->seats, link) {
+		seat->exclusive_client = NULL;
+	}
+
+	// redraw everything
```

Comment:
Maybe it's worth extracting this into a `desktop_damage_whole` function in `sway/desktop/desktop.c`?

Reply:
Since #6844 will end up reworking that anyway, I'd prefer to leave it as is unless you prefer it moved.

Reply:
Ack